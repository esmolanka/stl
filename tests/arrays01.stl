
type SomeArray a = exists (n : Nat). Int[n]

#eval SomeArray Float

#check forall a. SomeArray a <: forall b. SomeArray b

type SizedArray a = exists (n : Nat). { size : Nat n, elements : a[n] }

#eval SizedArray Int

#check forall a. SizedArray a <: forall b. SizedArray b

#eval exists (n : Nat). { ints : Int[n], doubles : Float[n] }

type Table (n : Nat) k v =
  { name : k[n]
  , value : v[n]
  }

type Maybe a = < Nothing | Just : a >

type Lookup = forall (n : Nat) k v. k -> Table n k v -> Maybe v

#eval Lookup

#check Lookup <: Lookup
