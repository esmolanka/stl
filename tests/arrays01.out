tests/arrays01.types:4:1-4:22:
  Normalised:
    ∃ (n : Nat). Int[n]
  Kind:
    Type
tests/arrays01.types:6:1-6:54:
  Subtype:
    ∀ a. ∃ (n : Nat). Int[n]
  Supertype:
    ∀ b. ∃ (n : Nat). Int[n]
  Result:
    OK
tests/arrays01.types:10:1-10:21:
  Normalised:
    ∃ (n : Nat). ∃ (ρ : Row). {size : Nat n, elements : Int[n] | ρ}
  Kind:
    Type
tests/arrays01.types:12:1-12:56:
  Subtype:
    ∀ a. ∃ (n : Nat). ∃ (ρ : Row). {size : Nat n, elements : a[n] | ρ}
  Supertype:
    ∀ b. ∃ (n : Nat). ∃ (ρ : Row). {size : Nat n, elements : b[n] | ρ}
  Result:
    OK
tests/arrays01.types:14:1-14:62:
  Normalised:
    ∃ (n : Nat). ∃ (ρ : Row). {ints : Int[n], doubles : Float[n] | ρ}
  Kind:
    Type
tests/arrays01.types:25:1-25:13:
  Normalised:
    ∀ (n : Nat). ∀ k. ∀ v.
      k
      -> (∃ (ρ : Row). {name : k[n], value : v[n] | ρ})
      -> ∀ (ρ : Row). <Nothing? : Unit | Just? : v | ... ρ>
  Kind:
    Type
tests/arrays01.types:27:1-27:24:
  Subtype:
    ∀ (n : Nat). ∀ k. ∀ v.
      k
      -> (∃ (ρ : Row). {name : k[n], value : v[n] | ρ})
      -> ∀ (ρ : Row). <Nothing? : Unit | Just? : v | ... ρ>
  Supertype:
    ∀ (n : Nat). ∀ k. ∀ v.
      k
      -> (∃ (ρ : Row). {name : k[n], value : v[n] | ρ})
      -> ∀ (ρ : Row). <Nothing? : Unit | Just? : v | ... ρ>
  Result:
    OK